Sources/ConsoleKit/Command/Base/CommandRunnable.swift

```diff
@@ -1,25 +1,73 @@
 import NIO
 
-/// Capable of being run on a `Console` using `Console.run(...)`.
-/// - note: This base protocol should not be used directly. Conform to `Command` or `CommandGroup` instead.
-public protocol CommandRunnable {
-    /// The supported options.
-    var options: [CommandOption] { get }
-
+/// A type-erased `CommandRunnable`.
+public protocol AnyCommandRunnable {
+    /// An instance of the type that represents the command's valid inputs/signature.
+    static var inputs: Inputs { get }
+    
     /// Text that will be displayed when `--help` is passed.
-    var help: [String] { get }
-
+    var help: String? { get }
+    
     /// The type of runnable. See `CommandRunnableType`.
     var type: CommandRunnableType { get }
+    
+    /// Runs the command against the supplied input.
+    func run(using context: AnyCommandContext) throws -> EventLoopFuture<Void>
+}
 
+/// Capable of being run on a `Console` using `Console.run(...)`.
+///
+/// - Note: This base protocol should not be used directly. Conform to `Command` or `CommandGroup` instead.
+public protocol CommandRunnable: AnyCommandRunnable {
+    /// The command's signiture.
+    ///
+    /// This type is made up of `Argument<T>` and `Option<T>` properties which
+    /// are the command's accepted arguments and options.
+    associatedtype Signature: Inputs
+    
+    /// An instance of the type that represents the command's valid signature.
+    ///
+    /// The type-specific implementation of `AnyCommandRunnable.inputs`.
+    static var signature: Signature { get }
+    
     /// Runs the command against the supplied input.
-    func run(using context: CommandContext) throws -> EventLoopFuture<Void>
+    func run(using context: CommandContext<Self>) throws -> EventLoopFuture<Void>
+}
+
+extension CommandRunnable {
+    
+    /// The default implementation of `AnyCommandRunnable.inputs`.
+    ///
+    /// - Returns: The `CommandRunnable.signature` value.
+    public static var inputs: Inputs {
+        return self.signature
+    }
+    
+    /// The default implementation of `CommandRunnable.signature`.
+    ///
+    /// - Returns: A new instance of `CommandRunnable.Signature`.
+    public static var signature: Signature {
```

Comment:
I think this should be a stored property so that we don't have to construct it each time. But given we won't be doing that _super_ often, it might not be a huge deal.

Reply:
The idea here is that we just allow a default implementation, and the user can override it because that property is required by the protocol. Most people won't care, but if you do, you can add the stored property to your command.